File: lib_ebook_db/src/main/java/com/ebook/db/entity/ChapterList.java
package com.ebook.db.entity;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import org.greenrobot.greendao.annotation.Entity;
import org.greenrobot.greendao.annotation.Generated;
import org.greenrobot.greendao.annotation.Id;
import org.greenrobot.greendao.annotation.Transient;

/**
 * 章节列表
 */
@Entity
public class ChapterList implements Parcelable, Cloneable {

    @Transient
    public static final Creator<ChapterList> CREATOR = new Creator<ChapterList>() {
        @Override
        public ChapterList createFromParcel(Parcel in) {
            return new ChapterList(in);
        }

        @Override
        public ChapterList[] newArray(int size) {
            return new ChapterList[size];
        }
    };
    private String noteUrl; //对应BookInfo noteUrl;
    private int durChapterIndex;  //当前章节数
    @Id
    private String durChapterUrl;  //当前章节对应的文章地址
    private String durChapterName;  //当前章节名称
    private String tag;
    private Boolean hasCache = false;
    @Transient
    private BookContent bookContent = new BookContent();

    protected ChapterList(Parcel in) {
        noteUrl = in.readString();
        durChapterIndex = in.readInt();
        durChapterUrl = in.readString();
        durChapterName = in.readString();
        tag = in.readString();
        bookContent = in.readParcelable(BookContent.class.getClassLoader());
        hasCache = in.readByte() != 0;
    }

    @Generated(hash = 1219340674)
    public ChapterList(String noteUrl, int durChapterIndex, String durChapterUrl,
                       String durChapterName, String tag, Boolean hasCache) {
        this.noteUrl = noteUrl;
        this.durChapterIndex = durChapterIndex;
        this.durChapterUrl = durChapterUrl;
        this.durChapterName = durChapterName;
        this.tag = tag;
        this.hasCache = hasCache;
    }

    @Generated(hash = 1539606642)
    public ChapterList() {
    }

    public static Creator<ChapterList> getCREATOR() {
        return CREATOR;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(noteUrl);
        dest.writeInt(durChapterIndex);
        dest.writeString(durChapterUrl);
        dest.writeString(durChapterName);
        dest.writeString(tag);
        dest.writeByte((byte) (hasCache == null ? 0 : hasCache ? 1 : 2));
        dest.writeParcelable(bookContent, flags);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public String getNoteUrl() {
        return this.noteUrl;
    }

    public void setNoteUrl(String noteUrl) {
        this.noteUrl = noteUrl;
    }

    public int getDurChapterIndex() {
        return this.durChapterIndex;
    }

    public void setDurChapterIndex(int durChapterIndex) {
        this.durChapterIndex = durChapterIndex;
    }

    public String getDurChapterUrl() {
        return this.durChapterUrl;
    }

    public void setDurChapterUrl(String durChapterUrl) {
        this.durChapterUrl = durChapterUrl;
    }

    public String getDurChapterName() {
        return this.durChapterName;
    }

    public void setDurChapterName(String durChapterName) {
        this.durChapterName = durChapterName;
    }

    public String getTag() {
        return this.tag;
    }

    public void setTag(String tag) {
        this.tag = tag;
    }

    public Boolean getHasCache() {
        return this.hasCache;
    }

    public void setHasCache(Boolean hasCache) {
        this.hasCache = hasCache;
    }

    public BookContent getBookContent() {
        return bookContent;
    }

    public void setBookContent(BookContent bookContent) {
        this.bookContent = bookContent;
    }

    @NonNull
    @Override
    protected Object clone() throws CloneNotSupportedException {
        ChapterList chapterList = (ChapterList) super.clone();
        chapterList.noteUrl = noteUrl;
        chapterList.durChapterUrl = durChapterUrl;
        chapterList.durChapterName = durChapterName;
        chapterList.tag = tag;
        chapterList.hasCache = hasCache;
        chapterList.bookContent = new BookContent();
        return chapterList;
    }

    @Override
    public String toString() {
        return "ChapterList{" +
                "noteUrl='" + noteUrl + '\'' +
                ", durChapterIndex=" + durChapterIndex +
                ", durChapterUrl='" + durChapterUrl + '\'' +
                ", durChapterName='" + durChapterName + '\'' +
                ", tag='" + tag + '\'' +
                ", hasCache=" + hasCache +
                ", bookContent=" + bookContent +
                '}';
    }
}

File: lib_ebook_db/src/main/java/com/ebook/db/entity/ChapterList.kt
package com.ebook.db.entity

import android.os.Parcelable
import com.ebook.db.ObjectBoxManager
import io.objectbox.annotation.Entity
import io.objectbox.annotation.Id
import io.objectbox.annotation.Unique
import io.objectbox.relation.ToOne
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize


/**
 * 章节列表
 */
@Parcelize
@Entity
data class ChapterList(
    /**
     * 对应BookInfo noteUrl;
     */
    var noteUrl: String = String(),
    /**
     * 当前章节数
     */
    @JvmField
    var durChapterIndex: Int = 0,
    /**
     * 当前章节对应的文章地址
     */
    @Unique
    var durChapterUrl: String = String(),
    /**
     * 当前章节名称
     */
    var durChapterName: String = String(),
    var tag: String = String(),
    var hasCache: Boolean = false,
    @Id var id: Long = 0
) : Parcelable {
    constructor(
        noteUrl: String,
        durChapterIndex: Int,
        durChapterUrl: String,
        durChapterName: String,
        tag: String,
        hasCache: Boolean
    ) : this() {
        this.noteUrl = noteUrl
        this.durChapterIndex = durChapterIndex
        this.durChapterUrl = durChapterUrl
        this.durChapterName = durChapterName
        this.tag = tag
        this.hasCache = hasCache
    }

    @IgnoredOnParcel
    lateinit var bookInfo: ToOne<BookInfo>

    @IgnoredOnParcel
    lateinit var bookContent: ToOne<BookContent>

    fun clone(): ChapterList {
        val chapterList = this.copy()
        ObjectBoxManager.chapterListBox.attach(chapterList)
        chapterList.bookContent = ToOne(chapterList, ChapterList_.bookContent)
        if (chapterList.bookContent.target == null) {
            if (this::bookContent.isInitialized && this.bookContent.target != null) {
                chapterList.bookContent.target = this.bookContent.target.clone()
            } else {
                // 尝试从数据库查找
                val bookContentFromDB = ObjectBoxManager.bookContentBox
                    .query(BookContent_.durChapterUrl.equal(this.durChapterUrl))
                    .build().findFirst()
                if (bookContentFromDB != null) {
                    chapterList.bookContent.target = bookContentFromDB
                }
            }
        }
        return chapterList
    }
}
